src/gadgets/curves/bn256/ec_mul.rs

```diff
@@ -0,0 +1,440 @@
+        let lambda = Self::u256_from_hex_str(cs, LAMBDA);
+        let b1 = Self::u256_from_hex_str(cs, B1);
+        let b2 = Self::u256_from_hex_str(cs, B2);
+        let g1 = Self::u256_from_hex_str(cs, G1);
+        let g2 = Self::u256_from_hex_str(cs, G2);
+
+        let k = convert_field_element_to_uint256(cs, scalar.clone());
+
+        // We take 8 non-zero limbs for the scalar (since it could be of any size), and 4 for G2
+        // (since it fits in 128 bits).
+        let g2_times_k = k.widening_mul(cs, &g2, 8, 4);
+        let c1 = g2_times_k.to_high();
+
+        // We take 8 non-zero limbs for the scalar (since it could be of any size), and 5 for G2
+        // (since it fits in 130 bits).
+        let g1_times_k = k.widening_mul(cs, &g1, 8, 5);
+        let c2 = g1_times_k.to_high();
+        
+        // Converting all constants to field elements
+        let mut b1 = convert_uint256_to_field_element(cs, &b1, scalar_field_params);
+        let mut b2 = convert_uint256_to_field_element(cs, &b2, scalar_field_params);
+        let mut c1 = convert_uint256_to_field_element(cs, &c1, scalar_field_params);
+        let mut c2 = convert_uint256_to_field_element(cs, &c2, scalar_field_params);
+        let mut lambda = convert_uint256_to_field_element(cs, &lambda, scalar_field_params);
+
+        // q1 <- c1 * b1
+        // q2 <- c2 * b2
+        let mut q1 = c1.mul(cs, &mut b1);
+        let mut q1 = q1.negated(cs);
+        let mut q2 = c2.mul(cs, &mut b2);
+        let mut q2 = q2.negated(cs);
+
+        // k2 <- q2 - q1
+        let mut k2 = q2.sub(cs, &mut q1);
+        k2.normalize(cs);
+
+        // k2_lambda <- k2 * lambda, k1 <- k - k2_lambda
+        let mut k2_times_lambda = k2.mul(cs, &mut lambda);
+        let mut k1 = scalar.sub(cs, &mut k2_times_lambda);
+        k1.normalize(cs);
+
+        let k1_u256 = convert_field_element_to_uint256(cs, k1.clone());
+        let k2_u256 = convert_field_element_to_uint256(cs, k2.clone());
+        
+        let low_pow_2_128 = pow_2_128.to_low();
+        
+        // Selecting between k1 and -k1 in Fq
+        let (_, k1_out_of_range) = k1_u256.overflowing_sub(cs, &low_pow_2_128);
+        let k1_negated = k1.negated(cs);
+        let k1 = <BN256ScalarNNField<F> as NonNativeField<F, BN256Fr>>::conditionally_select(
+            cs,
+            k1_out_of_range,
+            &k1,
+            &k1_negated,
+        );
+
+        // Selecting between k2 and -k2 in Fq
+        let (_, k2_out_of_range) = k2_u256.overflowing_sub(cs, &low_pow_2_128);
+        let k2_negated = k2.negated(cs);
+        let k2 = <BN256ScalarNNField<F> as NonNativeField<F, BN256Fr>>::conditionally_select(
+            cs,
+            k2_out_of_range,
+            &k2,
+            &k2_negated,
+        );
+
+        Self {
+            k1,
+            k2,
+            k1_was_negated: k1_out_of_range.negated(cs),
+            k2_was_negated: k2_out_of_range.negated(cs),
```

Comment:
the previous comment's update would also mean you need to remove the negation here